### PR TITLE
Documentation refactors

### PR DESCRIPTION
**Pull request by wct9667**
## Description

General refactors to update code consistency and fix some small errors.

- `config.json` Changed 2 numbers (`clientId` `devs`) to be strings to ensure `discord.js` works with them.
- `basicRemember.js` Uses the `apiCalls.js`. Works the same as before functionally.
- `print-storage` is now devOnly
### `remember-chunk.js` 
Parameters:

- `exclude-bot-messages` If bot messages should be excluded in the collection. Has a default stored in `config.json`

- `hours` Added an hours to remember: max 5.

- `minutes` Max minutes is now 59. 

Changed the structure of how the api is called to grab an array of messages, then loop through it to filter and save. Used a helper method called `getMessagesByTime`. Which gets messages and checks timestamps to loop until a given time has been hit. It also filters bot messages depending on the parameter.

### `remember-count.js` . 
Parameters:

- `exclude-bot-messages` If bot messages should be excluded in the collection. Has a default stored in `config.json`

Changed the structure of how the api is called to grab an array of messages, then loop through it to filter and save. Used a helper method called `getMessagesAndReturnID`. Which gets a specified number of messages, then loops through those messages to filter. It then returns the id of the last message so it can be used as a starting point for further message retrieval.

- `apiCalls.js` made a new function called `getNumberMessages`. This function gets a specified number of messages given the `channel` to retrieve from, the `numberToSave` messages to retrieve. Has an optional parameter for `id` to start from.

- `rememberedMessages.js` removed `saveNumberMessages` and `saveMessagesTime ` as their functionality was added to the commands or into `getNumberMessages`.

### mathUtils.js
Added a new `.js` file to handle math helper functions

Added `clamp` which clamps a given number to a min and a max.

Updated the `remember-count` and `remember-chunk` to use clamp.

## Issues: 
Was unable to fully make the `remember` files consistent. `remember-chunk` and `remember-count` use `discord.js`, while the others directly reference the API. I ran into problems with repeating results and number of request errors and determined to continue to use `discord.js` message fetching for the ones that have it. They now both save to the same format, so this presents no known issues in terms of getting the needed data.

### CODE REVIEW

File: src/commands/remember/basicRemember.js [hunk 1]
```diff
@@ -13,7 +14,7 @@ module.exports = {
             type: ApplicationCommandOptionType.String,
         },
         {
-            name: 'search-specific-channel',
+            name: 'channel-id',
             description: 'Searches the give channel for the message.',
```
Comment (BlckHawker) [inline src/commands/remember/basicRemember.js:18]: Should be changed to reference the id. Something like `The id of the channel that is being searched`

File: src/commands/remember/basicRemember.js [hunk 2]
```diff
@@ -35,12 +36,11 @@ module.exports = {
...
             if(specificChannelId){
                 //get the message from given channel
-                const channel = await client.channels.cache.get(specificChannelId);
-                msg =  await channel.messages.fetch(idMessage);
+                msg =  await getMessageObject(specificChannelId,idMessage);
             }
             else{
                 //get the message from current channel
-                msg =  await interaction.channel.messages.fetch(idMessage);
+                msg =  await getMessageObject(interaction.channel.id, idMessage);
```
Comment (BlckHawker) [inline src/commands/remember/basicRemember.js:43]: This can be shortened with the use of a ternary operator
```js
const specificChannelId = interaction.options.getString('search-specific-channel') ?? interaction.channel.id
msg =  await getMessageObject(specificChannelId,idMessage);
```

File: src/commands/remember/remember-chunk.js [hunk 3]
```diff
@@ -20,25 +62,38 @@ module.exports = {
...
+            const valueMinutes = interaction.options.get('minutes').value;
+
+            //bot messages
+            const excludeBotMessages = interaction.options.getBoolean('exclude-bot-messages') ?? defaultExcludeBotMessages;
 
             //validate accuracy
-            let accuracy = interaction.options.getNumber("speed") ?? 100;
+            let accuracy = interaction.options.getNumber("speed") ?? 50;
             if(accuracy < 25) accuracy = 25; 
             if(accuracy > 100) accuracy = 100;
```
Comment (BlckHawker) [inline src/commands/remember/remember-chunk.js:92]: If `speed` is directly related to `accuracy` (as it seems to be here), then it could be stated in the documentation that valid numbers are 25 - 100 inclusively

**Review by BlckHawker — COMMENTED** [review]
## PR Description Review
- Be sure to check for formatting issues such as `There is a bug with basicRemember I will fix tommorow` being on the same line as your Description header. The bug should probably be in the `Issues` header


~~- `remember-chunk.js has multiple changes` This can be worded better to give a more specific idea of what has been changed. A whole description is not needed for every minute detail, but a generalization would be nice. Whether that mean formatting via a linter, or reducing the amount of duplicated code. In this example, some of the details that could have been said were:
    - Refactored command to separate hours and minutes with the max being 5 and 59 respectively
    - Changed the speed of search to be 50
    - Add the ability to exclude bot messages~~

- I did not realize that there was more to the `remember-chunk.js has multiple changes` statement until after I wrote the statement above. I recommend if you are writing a lot about one specific topic to either use indented bulleted lists or use headers. Same applies with `remember-count.js has multiple changes.` Here's an example:

---
### remember-chunk.js

#### Parameters
- `exclude-bot-messages` If bot messages should be excluded in the collection. Has a default stored in config.json
- `hours` Added an hours to remember: max 5.
- `minutes` Max minutes is now 59.
#### Notes
- Changed the structure of how the api is called to grab an array of messages, then loop through it to filter and save. Used a helper method called getMessagesByTime. Which gets messages and checks timestamps to loop until a given time has been hit. It also filters bot messages depending on the parameter.
---


Overall, your descriptions have a good amount of detail to understand what you did in the PR, but the formatting can make that information hard to follow. You don't need to follow my guidelines exactly, but I highly recommend having data that is broken up via line breaks linked in some way
## PR Review

Everything looks good so far. I made some suggestions that are not necessary, but could be beneficial. I can review again once the bug with `basicRemember` is fixed

File: src/commands/remember/remember-count.js [hunk 4]
```diff
@@ -1,5 +1,42 @@
...
+        startId = m[1].id;
+
+        // exlude bot messages if option is enabled
+        if (excludeBotMessages && m[1].author.bot) {
+            continue;
+        }
+
+        const message = parseMessage(m[1])
+        messagesToSave.push(message)
+    }
```
Comment (tstodter) [inline src/commands/remember/remember-count.js:36]: This should be an equivalent alternative
```js
messagesToSave = messageObjArray
    .filter(([_, msg]) => !(excludeBotMessages && msg.author.bot))
    .map(([_, msg]) => parseMessage(msg));
startId = messageObjArray.at(-1)[1].id; // last element
```

File: src/commands/remember/remember-count.js [hunk 5]
```diff
@@ -1,5 +1,42 @@
...
+const getMessagesAndReturnId = async(messagesToSave, channel, num, excludeBotMessages, startId) =>{
+    let messageObjArray;
+    if(startId){
+         //get the first 100 messages at a specifc point
+         messageObjArray = await getNumberMessages(channel, num, startId)
+    }
+    else{
+        //get the num messages starting with the latest one
+        messageObjArray = await getNumberMessages(channel, num)
+    }
```
Comment (tstodter) [inline src/commands/remember/remember-count.js:15]:
```suggestion
    messageObjArray = await getNumberMessages(channel, num, startId);
```
This should be the same thing; if there's no startId, it's `undefined` anyway.

File: src/utils/apiCalls.js [hunk 6]
```diff
@@ -42,9 +42,23 @@ const getMessagesAfterId = async (channelId, limit, afterId, addFirstMessage = f
...
+const getNumberMessages = async(channel, numberToSave, id) =>{
+    let messages;
+
+    //if id begin there
+    if (id) 
+        messages = await channel.messages.fetch({ cache: false, limit: numberToSave, before: id });
+    else
+        messages = await channel.messages.fetch({ cache: false, limit: numberToSave});
+            
+    return messages;
```
Comment (tstodter) [inline src/utils/apiCalls.js:55]: Alternative I generally prefer in my own style:
```js
return id
    ? channel.messages.fetch({cache: false, limit: numberToSave, before: id})
    : channel.messages.fetch({cache: false, limit: numberToSave});
```
You also don't really need async/await here, as you're just returning the result of this `.fetch()`, which is itself a promise. So my suggestion is equivalent to your code.

File: src/commands/remember/remember-chunk.js [hunk 7]
```diff
@@ -20,25 +62,38 @@ module.exports = {
...
 
-            const numberOfMinutes = (value > 600 ? 600 : value).toString() + 'm'; //formatting for ms
+            //format and clamp
+            const numberOfHours = (valueHours > 5 ? 5 : valueHours).toString() + 'h'; //formatting for ms
+            const numberOfMinutes = (valueMinutes > 59 ? 59 : valueMinutes).toString() + 'm'; //formatting for ms
```
Comment (tstodter) [inline src/commands/remember/remember-chunk.js:96]: For the clamping above you could also do something like `Math.min(valueMinutes, 59)` or `accuracy = Math.min(accuracy, 100)`.

**Review by tstodter — APPROVED** [review]
LGTM, untested

File: src/commands/remember/remember-chunk.js [hunk 8]
```diff
@@ -20,25 +61,38 @@ module.exports = {
...
+            //bot messages
+            const excludeBotMessages = interaction.options.getBoolean('exclude-bot-messages') ?? defaultExcludeBotMessages;
 
             //validate accuracy
-            let accuracy = interaction.options.getNumber("speed") ?? 100;
-            if(accuracy < 25) accuracy = 25; 
-            if(accuracy > 100) accuracy = 100;
+            let accuracy = interaction.options.getNumber("speed") ?? 50;
+            accuracy = Math.max (accuracy, 25); 
+            accuracy = Math.min(accuracy, 100);
```
Comment (tstodter) [inline src/commands/remember/remember-chunk.js:91]: I wasn't paying attention in my initial review :/ No need to change anything here, just a tip. This process is usually called "clamp", so I'll sometimes do something like this:
```js
const clamp = (min, max, num) => Math.max(min, Math.min(num, max));
let accuracy = clamp(25, 100, interaction.options.getNumber("speed") ?? 50);
```
Benefit is that it's self-documenting, and (I think at least) helps your mind abstract over minute details. Plus you're actually clamping several times in this file, so I might recommend making clamp a global in this file and just using throughout.

File: src/commands/remember/remember-count.js (hunk 4 above)
Comment (tstodter) [reply src/commands/remember/remember-count.js:28]: To be clear, I strongly prefer data-oriented approaches like this (where everything basically becomes array processing, transforming and filtering and modifying arrays into arrays into arrays). My suggestion is definitely denser, which may come at a cost of readability, so if I suggest an "alternative" I'm hoping you'll just do whatever you think works best (take or leave my approach).

Comment (wct9667) [reply src/commands/remember/remember-count.js:28]: The main reason I didn't implement this was the messageObjArray wasn't iterable normally, both filter and map threw errors. And I couldn't get startId the way you suggested as id was marked as undefined. I was in a rush cause of my car appointment so I can take a closer look later.

Comment (tstodter) [reply src/commands/remember/remember-count.js:28]: Could try `[...messageObjArray]`
